### PR TITLE
Teleprompt Base Class -> ABC, Some Static Typing, and Ruff Fixes/Updates

### PR DESCRIPTION
**Pull request by mgbvox**
Only clear dsp.setting.trace in assertions.py if a trace object exists. 

Adapt YouRM API call signature to match the documentation on https://api.you.com/api-key; update return data to respect self.k (as opposed to returning all snippets from the first K hits).

### CODE REVIEW

**Comment by insop** [comment]
It's be great if you add a you.com entry to this document? that will help others who want to use you.com
If you prefer, a separate PR would work for sure.

**Comment by mgbvox** [comment]
@insop do you mean to the documentation? Sure, I can add that. Will get to it this evening.

File: dspy/primitives/assertions.py [hunk 1]
```diff
@@ -237,7 +237,9 @@ def wrapper(*args, **kwargs):
                     break
                 else:
                     try:
-                        dsp.settings.trace.clear()
+                        # only try to clear the trace if one exists
```
Comment (okhat) [inline dspy/primitives/assertions.py:240]: Overall this makes sense but I worry it's hiding a different issue under the hood?

cc @arnavsinghvi11 @manishshettym @Shangyint

**Comment by mgbvox** [comment]
Had less bandwidth than I anticipated, I plan to sit down and contribute a bunch come early March.

**Comment by okhat** [comment]
Tagging @arnavsinghvi11 for the issue above it comes up a lot: dsp.settings.trace being None by default causes issues. Is setting it globally to [] correct? Probably not? It affects optimization somewhere? Is there any code checking for dspy.settings.trace that assumes it None and [] are different?

**Comment by arnavsinghvi11** [comment]
> dsp.settings.trace

Yeah I believe we should set it to [] going forward since there are [checks in assertion.py](https://github.com/stanfordnlp/dspy/issues/432). It seems that the optimizations logic also set [trace = [] ](https://github.com/stanfordnlp/dspy/blob/b2816c4a35e3144a06752423de1afb8e68e1005f/dspy/teleprompt/bootstrap.py#L133) internally so I believe we can make this change globally in settings. Let me know if that aligns and I can push a PR for it.

**Pull request closed without merging**